src/backend/distributed/connection/connection_management.c

```diff
@@ -429,16 +429,6 @@ FindAvailableConnection(dlist_head *connections, uint32 flags)
 			continue;
 		}
 
-		if (connection->forceCloseAtTransactionEnd)
-		{
-			/*
-			 * This is a connection that should be closed, probabably because
-			 * of old connection options. So we ignore it. It will
-			 * automatically be closed at the end of the transaction.
```

Comment:
I guess the proper change would be the following. If we have already sent the `BEGIN` over the connection, we should always give that connection. Otherwise, we can skip. Please improve the comments 
```C
		if (connection->forceCloseAtTransactionEnd && !connection->remoteTransaction.beginSent)
		{
			/*
			 * TODO: improve comment. if begin has been sent, we have to use the same connection, else skip.
			 * This is a connection that should be closed, probabably because
			 * of old connection options. So we ignore it. It will
			 * automatically be closed at the end of the transaction.
			 */
			continue;
		}
```